sphinx/util/inspect.py

```diff
@@ -241,14 +239,58 @@ def isclassmethod(
     return False
 
 
+def is_classmethod_descriptor(
+    obj: Any, cls: Any = None, name: str | None = None
+) -> TypeIs[types.ClassMethodDescriptorType]:
+    """Check if the object is a :class:`~types.ClassMethodDescriptorType`.
+
+    This is typically useful to check if a built-in method is a class method.
+    """
+    if isinstance(obj, types.ClassMethodDescriptorType):
+        return True
+    if cls and name:
+        # trace __mro__ if the method is defined in parent class
+        sentinel = object()
+        for basecls in getmro(cls):
+            meth = basecls.__dict__.get(name, sentinel)
+            if meth is not sentinel:
+                return isinstance(meth, types.ClassMethodDescriptorType)
+    return False
+
+
+def is_builtin_classmethod(obj: Any, cls: Any = None, name: str | None = None) -> bool:
+    """Check if the object is a class method implemented in C."""
+    if is_classmethod_descriptor(obj, cls, name):
+        return True
+    if (
+        isbuiltin(obj)
+        and getattr(obj, '__self__', None) is not None
+        and isclass(obj.__self__)
+    ):
+        return True
+    if cls and name:
+        # trace __mro__ if the method is defined in parent class
+        sentinel = object()
+        for basecls in getmro(cls):
+            meth = basecls.__dict__.get(name, sentinel)
+            if meth is not sentinel:
+                return is_builtin_classmethod(meth)
```

Comment:
Does this still need to recurse? I'd prefer being explicit here (even if duplicative):

```suggestion
                return is_classmethod_descriptor(obj, cls, name) or (
                    isbuiltin(obj)
                    and getattr(obj, '__self__', None) is not None
                    and isclass(obj.__self__)
                )
```

Reply:
I aligned with the previous style for is_classmethod but I can also inkine the patch.